source/elements/oneDAL/include/oneapi/dal/array.hpp

```diff
@@ -0,0 +1,290 @@
+namespace oneapi::dal {
+
+/// @tparam Data The type of the memory block elements within the array.
+///
+/// @pre $Data$ cannot be const-qualified.
+template <typename Data>
+class array {
+public:
+    using data_t = Data;
+
+public:
+    /// Allocates a new memory block for mutable data, does not initialize it,
+    /// creates a new array instance by passing a pointer to memory block.
+    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    ///
+    /// @param queue The SYCL* queue object.
+    /// @param count The number of elements of type $Data$ to allocate.
+    /// @param alloc The kind of USM to be allocated.
+    /// @pre :expr:`count > 0`
+    static array<Data> empty(const sycl::queue& queue,
+                             std::int64_t count,
+                             const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
+
+    /// Allocates a new memory block for mutable data, fills it with a scalar value,
+    /// creates a new array instance by passing a pointer to memory block.
+    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    ///
+    /// @tparam Element The type from which array elements of type $Data$ can be constructed.
+    ///
+    /// @param queue   The SYCL* queue object.
+    /// @param count   The number of elements of type $Data$ to allocate.
+    /// @param element The value that is used to fill a memory block.
+    /// @param alloc   The kind of USM to be allocated.
+    /// @pre :expr:`count > 0`
+    /// @pre Elements of type ``Data`` are constructible from ``Element`` type.
+    template <typename Element>
+    static array<Data> full(sycl::queue& queue,
+                            std::int64_t count,
+                            Element&& element,
+                            const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
+
+    /// Allocates a new memory block on mutable data, fills it with zeros,
+    /// creates a new array instance by passing a pointer to memory block.
+    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    ///
+    /// @param queue   The SYCL* queue object.
+    /// @param count   The number of elements of type $Data$ to allocate.
+    /// @param alloc   The kind of USM to be allocated.
+    /// @pre :expr:`count > 0`
+    static array<Data> zeros(sycl::queue& queue,
+                             std::int64_t count,
+                             const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
+
+    /// Creates a new array instance by passing the pointer to externally-allocated memory block
+    /// for mutable data. It is responsibility of the calling application to free memory block
+    /// as the array shall not free it when reference counter equals to zero.
```

Comment:
I don't think you can say "reference counter equals to zero", because the counter is a program that counts, not an object that holds the results. 

```suggestion
    /// for mutable data. It is responsible for calling the application to free the memory block
    /// as the array shall not free it when the reference count is zero.
```

applies in other places as well